scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,219 @@
+(ns cook.kubernetes.controller
+  (:require [clojure.tools.logging :as log]
+            [cook.datomic :as datomic]
+            [cook.kubernetes.api :as api]
+            [cook.mesos.sandbox :as sandbox]
+            [cook.scheduler.scheduler :as scheduler])
+  (:import (clojure.lang IAtom)
+           (io.kubernetes.client.models V1Pod V1ContainerStatus V1PodStatus)))
+
+;
+;   Wire up a store with the results.
+;
+
+(def lock-object (Object.))
+(defn calculate-lock
+  "Given a pod-name, return an object suitable for locking for accessing it."
+  [_]
+  ; TODO: Should do lock sharding based on hash of pod-name.
+  lock-object)
+
+(defn expected-state-equivalent?
+  "Is the old and new state equivalent?"
+  [old-state new-state]
+  (= old-state new-state); TODO
+  )
+
+(defn existing-state-equivalent?
+  "Is the old and new state equivalent?"
+  [old-state new-state]
+  (= old-state new-state) ; TODO. This is not right. We need to tune/tweak this to suppress otherwise identical states so we don't spam.
+  )
+
+(defn remove-finalization-if-set-and-delete
+  [api-client expected-state-dict pod]
+  (api/remove-finalization-if-set-and-delete api-client pod)
+  expected-state-dict)
+
+(defn kill-task
+  [api-client expected-state-dict pod]
+  (api/kill-task api-client pod)
+  expected-state-dict)
+
+(defn launch-task
+  [api-client expected-state-dict]
+  (api/launch-task api-client expected-state-dict)
+  expected-state-dict)
+
+(defn update-or-delete!
+  "Given a map atom, key, and value, if the value is not nil, set the key to the value in the map. If the value is nil, delete the key entirely."
+  [^IAtom map-atom key value]
+  (if (nil? value)
+    (swap! map-atom dissoc key)
+    (swap! map-atom assoc key value)))
+
+(defn container-status->failure-reason
+  "Maps kubernetes failure reasons to cook failure reasons"
+  [^V1PodStatus pod-status ^V1ContainerStatus status]
+  ; TODO map additional kubernetes failure reasons
+  (let [terminated (-> status .getState .getTerminated)]
+    (cond
+      (= "OutOfMemory" (.getReason pod-status)) :reason-container-limitation-memory
+      (= "Error" (.getReason terminated)) :reason-command-executor-failed
+      (= "OOMKilled" (.getReason terminated)) :reason-container-limitation-memory
+      :default (do
+                 (log/warn "Unable to determine kubernetes state for pod" {:pod-status pod-status
+                                                                           :container-status status})
+                 :unknown))))
+
+(defn handle-status-update
+  "Helper function for calling scheduler/handle-status-update"
+  [kcc mesos-status]
+  (scheduler/write-status-to-datomic datomic/conn
+                                     @(:pool->fenzo-atom kcc)
+                                     mesos-status))
+
+(defn- get-job-container-status
```

Comment:
Move this into API, and rename 'get-pod-status-for-cook-container-in-pod'. Also, why does this exist? Shouldn't we be using the synthesized-state instead?

Reply:
My vote would be to keep it here since it's used here, and it's private. We use this because we need the actual state object to determine the failure reason.

Reply:
Pasted into docs. Thank you.